lib/spack/llnl/util/filesystem.py

```diff
@@ -539,6 +539,28 @@ def get_filetype(path_name):
     return output.strip()
 
 
+def is_executable_nonsymlink_script(path):
+    """
+    Returns whether the path is an executable script with a shebang.
+    Return False when the path is a *symlink* to an executable script.
+    """
+    try:
+        st = os.lstat(path)
+        # Should not be a symlink
+        if stat.S_ISLNK(st.st_mode):
+            return False
+
+        # Should be executable
+        if not st.st_mode & (stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH):
+            return False
+
+        # Should start with a shebang
+        with open(path, 'rb') as f:
+            return f.read(2) == b'#!'
```

Comment:
This is more-targeted: it would ignore scripts that don't have a shebang line. The original intent was primarily to filter the shebang, so this change seems fine, but I think a more specific name for this function would be helpful.

Reply:
That makes no sense? `file x.py` returns ASCII text, not "script".

Reply:
In short
- `file` calls "(executable) text files without shebang" `ascii text`, meaning, we did not handle those files as "scripts".
- `file` calls "executable and non-executable text files with shebang" `script`
- linux handles executable files with shebang in binfmt_script.c

So it seems there is some level of agreement that scripts are files with a shebang

Reply:
That's interesting. I think it's still worth documenting explicitly: since we're moving away from "whatever `file` thinks is a script" to our own implementation (which - perhaps like the `file` command - does not identify certain scripts as such). Do you know for example if this would apply to bash scripts without shebang lines that are included as part of the python package installation (i.e. does `file` recognize a bash script without a shebang line as a script)?

Reply:
Well, if you have no shebang, how would you possibly know the interpreter should be bash instead fish or sh? So I think `ascii text` is indeed the only sensible file type.

Reply:
FWIW, this is the magic database for scripts: https://github.com/file/file/blob/master/magic/Magdir/varied.script

There's little of real interest there since I don't think we care what kind of script it is, the main possible pitfall is for something that gets a *higher* strength than the prefix shebang and gets handled some other way as a result.  Honestly I think this is fine for this purpose, it's already got some other information going into the selection before it gets here.

Reply:
Thanks for the pointer, I wanted to check that out but forgot about it.